packages/react-form/src/hooks/dirty.ts

```diff
@@ -1,14 +1,6 @@
-import {useMemo} from 'react';
-
 import {FieldBag} from '../types';
-import {fieldsToArray} from '../utilities';
+import {reduceFields} from '../utilities';
 
 export function useDirty(fieldBag: FieldBag) {
-  const fields = fieldsToArray(fieldBag);
-
-  return useMemo(
-    () => fields.some(field => field.dirty),
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    fields,
-  );
+  return reduceFields(fieldBag, (dirty, field) => dirty || field.dirty, false);
```

Comment:
The problem with having `fields` as I had it is that `useList` fields are dynamic which causes improper memoization of fields (in the second argument of) `useMemo`.

We could alternately go back to using the more static `Object.values(fields)` where `fields` are those declared in `useForm` but it doesn't properly memoize nested fields or lists, making `useMemo` useless.

I would argue that removing hooks altogether would have little to no performance difference as shown here. In fact, removing the need for `useMemo` plus not needing to create a new array to watch is arguably more performant. In the long term, I would recommend deprecating `useDirty` and instead replace it with a utility function.